### DISCUSSION TITLE
How does the coordinate system work?

### DISCUSSION THREAD

**Discussion by wowtah** [xyflow/xyflow#4311] [Q&A] [answered]
Can somebody explain the coordinate system of the canvas to me? Especially in relation to node positions. I can't seem to wrap my head around it. 

Let's take the ViewportLogger from the Devtools for instance (https://reactflow.dev/learn/advanced-use/devtools-and-debugging#viewport-logger).

As you can see in this screenshot

![image](https://github.com/xyflow/xyflow/assets/30867746/b30c83bd-7654-4dee-b280-672ba05b2afd)

the viewport coordinates show as: 
x: -243.90
y: -4.40 

**Question 1: Does this mean the top and left of the visible canvas are at these coordinates? Or do they mark another place?**

Let's assume for now they are indeed the top left of the canvas, and let's continue... 

For the sake of the example, lets correct these coordinates with the zoom of 0.98 (I couldn't get the zoom at 1.0 in the screenshot), so let's say we are around x: -250 and y: -5. 

Now, if you check the position of the top-left node in the screenshot (which I place all the way to the top left edge of the visible area), the node shows it lives at position: x: 250, y: 5. These numbers seem very familiar... they are the positive versions of the viewport coordinates we are seeing. 

**Question 2: How does this work? If the viewport x/y is indeed marking the top left corner, then I would assume the position of the node and the coordinates are the same? Why this negative version of them?**

Allright, couple more things. 

If I drag the viewport to the right, more of the (previously invisible) left side of the canvas comes into view. This makes sense. However, the viewport coordinates seem to go in the other direction. If I drag to the right, my x coordinate goes up, instead of down. Same with dragging up and down. Coordinates seem to go in the opposite direction.  

**Question 3: Are the viewport coordinates in reverse? If so, why?**

Okay, sorry for the long text, but one more thing that I don't understand. 

If (indeed) the viewport coordinates show the top left of the visible part of the canvas, then I would expect this value to be the same whenever I align my top left at the same position, regardless of the zoom level. So when I place my node at a position, and I drag the viewport so that the node is in the top left, then regardless of the zoom level, I would expect the viewport coordinates to show the same position, since they should mark the position on the canvas? As you can see in this second screenshot (where I zoomed in to 2.00), the coordinates have changed. 

![image](https://github.com/xyflow/xyflow/assets/30867746/7c115170-ff5e-44e8-9a20-27bcdecd43bd)

If I would do some funky math (remove the negative of the viewport coords and then divide by the zoom level), I can reach the position x/y of the top-left node again (250 / 5). But before blindly using funky math like this, I really want to understand how this works. 

**So final question (which may be the only one that needs answering actually): how to understand these concepts correctly?** 

I am using:
`"reactflow": "^11.11.3",`

TL;DR: how does the viewport coordinate system correlates to node positions?

**Comment by bcakmakoglu** [comment]
<img width="737" alt="image" src="https://github.com/xyflow/xyflow/assets/78412429/87350fde-127e-41ec-b415-65db84899465">

As you can see in the image above, when the viewport is untouched the top-left coordinate is 0,0.

All the other questions you made kinda are based on the initial assumption that that wouldn't be the case because you read some numbers on the graph that didn't *seem* to match (they do ^^)

**Reply by bcakmakoglu** [reply]
Although maybe you confusion is about this?
If you start dragging the pane to the left the viewport x,y values go *down*.
If you start dragging the pane to the right the viewport x,y values go *up*.

That does sort of feel like it's "reverse" since -100,-100 on a node would put it further left and up while a viewport coordinate of 100, 100 would put the viewport further right and down ^^

**Reply by wowtah** [reply]
> That does sort of feel like it's "reverse" since -100,-100 on a node would put it further left and up while a viewport coordinate of 100, 100 would put the viewport further right and down ^^

Exactly, why is this? At 0:0 everything lines up. But move your node to 200:200 and align with the top left again. Now the viewport is at -200 / -200. Why the difference?

I would expect the canvas to work like this?
![image](https://github.com/xyflow/xyflow/assets/30867746/8e7dfd43-f6b3-4b0f-a4cb-981a09f83494)

**Reply by bcakmakoglu** [reply, marked as answer]
Well, the x,y values of the viewport aren't *technically* coordinates (that's why it's stored internally as `transform`), they are values used to apply a transformation to the viewport.
If we were to reverse these numbers, the viewport transformation would be reversed.
Basically, we take the viewport container and transform it by x and y, so an x value of 300 would *push* the transformed container 300px to the *right* while the visible part would seem to be going *left*.
Same for the y value of course.

In reality it's the container with the elements that is moved around with these values not the "background" (viewport) itself.

https://github.com/xyflow/xyflow/assets/78412429/e45be181-34e0-47cf-b951-20bb4573df5a

**Reply by wowtah** [reply]
okay, it makes sense if the values are transform values and not grid coordinates. This is why they feel 'reversed'. Thanks for the explanation. 

So, if I want to calculate the viewport coordinates that align with how Node positions work, then I would indeed do something like:

`invert the coordinate (plus to minus and vice versa) and divide by zoom level. `

```
export default function ViewportLogger() {
  const { [0]: x, [1]: y, [2]: zoom } = useStore((s) => s.transform);

  // inverse the x and y values. So positive becones negative and vice versa
  const { new_x, new_y } = { new_x: -x, new_y: -y };

  // divive new_x and new_y by the zoom level
  const { final_x, final_y } = { final_x: new_x / zoom, final_y: new_y / zoom };

  const viewport = `x: ${final_x.toFixed(2)}, y: ${final_y.toFixed(2)}, zoom: ${zoom.toFixed(2)}`;

  return (
    <Panel position="top-right" className="pt-4">
      {viewport}
    </Panel>
  );
}
```

(put here in separate steps for clarity). 

This way the Node positioning aligns with the canvas coordinates.

**Reply by bcakmakoglu** [reply]
Yeah, if you want to think of it like that then that works :)

**Reply by wowtah** [reply]
> Yeah, if you want to think of it like that then that works :)

Yeah, for my users, I think it makes more sense if the viewport coordinates show where they are looking on the canvas, instead of 'how should the viewport be transformed in order to calculate where we are looking'. 

But makes sense from a coding standpoint to work with the transform values.
